### ISSUE TITLE
improve functional tests (stability and easy to debug)

### ISSUE THREAD

**Issue by xukai92** [instructlab/instructlab#748]
the current tests in Bash scripts seem to be flaky and hard to debug.
we need figure out a better solution and migrate to that.

**Comment by anik120** [comment]
Fwiw, I don't think this is achievable in a week. We're talking about the rewrite of an entire e2e test suite here. Probably needs to have a proper implementation plan in place before we can schedule this for delivery and commit to it.

**Comment by nathan-weinberg** [comment]
@alimaredia do you still intend to work on this?

**Comment by alimaredia** [comment]
Yes, we want to improve the functional tests for the 0.19.0 milestone.

**Comment by nathan-weinberg** [comment]
I've identified the functionalities of the functional test script as it stands today

## test functions
We have **15** functional tests currently, I've grouped them into six main categories

### server
- test_bind_port
- test_temp_server
- test_temp_server_sigint
- test_temp_server_ignore_internal_messages
- test_server_welcome_message
- test_server_template_value
- test_server_chat_template

### model train
- test_model_train

### model download
- test_oci_model_download_with_vllm_backend

### model test
- test_model_test

### data generate
- test_data_generate

### logging
- test_model_print
- test_ctx_size
- test_loading_session_history
- test_no_chat_logs

## utility functions

We have **4** functional tests currently
- cleanup
- init_test_script
- to_upper
- wait_for_server

I've opened https://github.com/instructlab/instructlab/pull/2248 to clean up the script - not a resolution to this issue, but will move us in the right direction

**Comment by courtneypacheco** [comment]
Hey @xukai92! Just following up since this issue is still open...

Are you satisfied with the current state of our functional test script? If not, are there specific areas of the file that you still find confusing or difficult to debug?
